logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/backup/CreateBackupUseCase.kt

```diff
@@ -78,32 +78,33 @@ internal class CreateBackupUseCaseImpl(
     private val idMapper: IdMapper = MapperProvider.idMapper(),
 ) : CreateBackupUseCase {
 
-    override suspend operator fun invoke(password: String): CreateBackupResult = withContext(dispatchers.default) {
-        val userHandle = userRepository.getSelfUser()?.handle?.replace(".", "-")
-        val timeStamp = DateTimeUtil.currentSimpleDateTimeString()
-        val backupName = createBackupFileName(userHandle, timeStamp)
-        val backupFilePath = kaliumFileSystem.tempFilePath(backupName)
-        deletePreviousBackupFiles(backupFilePath)
-
-        val plainDBPath =
-            databaseExporter.exportToPlainDB(securityHelper.userDBOrSecretNull(userId))?.toPath()
-                ?: return@withContext CreateBackupResult.Failure(StorageFailure.DataNotFound)
-
-        try {
-            createBackupFile(userId, plainDBPath, backupFilePath).fold(
-                { error -> CreateBackupResult.Failure(error) },
-                { (backupFilePath, backupSize) ->
-                    val isBackupEncrypted = password.isNotEmpty()
-                    if (isBackupEncrypted) {
-                        encryptAndCompressFile(backupFilePath, password)
-                    } else CreateBackupResult.Success(backupFilePath, backupSize, backupFilePath.name)
-                })
-        } finally {
-            databaseExporter.deleteBackupDBFile()
+    override suspend operator fun invoke(password: String, onProgress: (Float) -> Unit): CreateBackupResult =
```

Comment:
I guess a callback works, but it could be a bit safer to instead return a `Flow` that emits progress.

A Flow can be configured to handle back pressure more gracefully. For example, what if the backup creation calls `onProgress` faster than it can be executed?

I think it would be nicer to just return a `Flow<BackupProgress>` instead.
And `BackupProgress` could be something like:

```kotlin
sealed interface BackupProgress {
    data class Complete(val result: CreateBackupResult): BackupProgress
    data class Ongoing(val progress: Float): BackupProgress
}
```

-----

Another argument for Flow:
if we change the implementation of `CreateBackupUseCase` and it runs on a different scope, this different scope would keep a reference to `onProgress`, which could lead to a leak.

It's easier to just return a flow and if it is cancelled, it is cancelled :)

Reply:
My initial implementation, was done using a flow. I need to ask about the iOS implementation though. The return of the actual function is Either, so changing that would require to do the same functionality in iOS. 

I know that is the best approach and for sure will try to implement it. 

Thanks for the input man

Reply:
Amazing, then i will do the refactoring of this using flow 👍🏻